src/find_service/server.rs

```diff
@@ -15,73 +12,54 @@ fn convert_system_time_error(time_error: time::SystemTimeError) -> io::Error {
     io::Error::new(io::ErrorKind::Other, time_error)
 }
 
-type PublisherStore = Arc<RwLock<HashMap<String, proto::Registration>>>;
+#[derive(Debug, Clone)]
+pub struct PublisherNotFoundError;
+
+impl fmt::Display for PublisherNotFoundError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "invalid first item to double")
+    }
+}
+
+pub trait PublisherStore: Send + Sync + Clone + 'static {
+    fn insert_publisher(&self, publisher_name: &String, registration: proto::Registration);
```

Comment:
Prefer `&str` over `&String` because it's more flexible and deref coercion will automatically turn a `&String` into a `&str`.

```suggestion
    fn insert_publisher(&self, publisher_name: &str, registration: proto::Registration);
```